### ISSUE TITLE
Download count

### ISSUE THREAD

**Issue by Ionys320** [DirectoryLister/DirectoryLister#399]
Hello,
Could you please add a download count for each file and the total of downloads for a directory, visible directly in the files list (maybe between the size and the date)?

Thanks!

**Comment by xawlw** [comment]
+1 it would be cool to have that

**Comment by wloot** [comment]
Sorry but is there any priority to implement this feature  @PHLAK

**Comment by PHLAK** [comment]
No, this is not a priority. I haven't been able to come up with a way to implement this that satisfies my desires to keep things simple and not require configuration out of the box.

**Comment by xawlw** [comment]
> No, this is not a priority. I haven't been able to come up with a way to implement this that satisfies my desires to keep things simple and not require configuration out of the box.

I don't think it's possible to make without using a database.

**Comment by PHLAK** [comment]
@xawlw You could always store the data in flat files. My current best thought is to use an sqlite database (file) but that's still technically a database (though probably easier for people to set up).

**Comment by xawlw** [comment]
> @xawlw You could always store the data in flat files. My current best thought is to use an sqlite database (file) but that's still technically a database (though probably easier for people to set up).

Yes I think storing it in a sqlite file would be the best idea

**Comment by PHLAK** [comment]
The other major obstacle with this feature is having to process the links with PHP instead of serving the files directly. This adds additional development overhead as well as a possible (though probably insignificant) performance hit.

**Comment by Ionys320** [comment]
I created an API to work with the DirectoryLister, and I can tell you that the performance is not impacted ;-)

**Comment by xawlw** [comment]
> The other major obstacle with this feature is having to process the links with PHP (to record the download) instead of serving the files directly. This adds additional development overhead as well as a possible (though probably insignificant) performance hit.

You can make it optional, Also processing the link via php would be better for me as I want to add countdown timer to start the download

**Comment by xawlw** [comment]
> I created an API to work with the DirectoryLister, and I can tell you that the performance is not impacted ;-)

Hi, can you share the API, please?

**Comment by AndroiableDroid** [comment]
@PHLAK Plz review it.

**Comment by Massi-X** [comment]
Just a quick idea: you can use [xattr_set](https://www.php.net/manual/en/function.xattr-set.php) to store the download count directly on the file itself instead of relying on external storage. Obviously managing the link through PHP is still necessary but I think it is a fair compromise (I wouldn't be to worry about performance). If I have some time I will look into an implementation.

**Comment by Massi-X** [comment]
Besides what said above, I have improved and bugfixed your code. Feel free to use it if you want

File: FileDownloads.php
```
<?php

namespace App\ViewFunctions;

use Symfony\Component\Finder\SplFileInfo;

function updateDownloads(): void
{
	$dir = getcwd() . '/app/.downloads/.';
	$filename = str_replace(DIRECTORY_SEPARATOR, '', $_GET['download']);
	$handle = fopen($dir . $filename, 'r+');

	if (!$handle)
		return;

	$data = (int) fread($handle, (int) filesize($dir . $filename));
	fclose($handle);

	$handle = fopen($dir . $filename, 'w');
	if (!$handle)
		return;

	++$data;
	fwrite($handle, strval($data));
	fclose($handle);

	header('Location: ' . $filename);
	exit();
}

if (isset($_GET['download'])) {
	updateDownloads();
}

class FileDownloads extends ViewFunction
{
	/** @var string The function name */
	protected $name = 'file_downloads';

	/** @var string Dir name */
	private $dir;

	public function __construct()
	{
		$this->dir = getcwd() . '/app/.downloads/.';
	}

	/** Get the modified time from a file object. */
	public function __invoke(SplFileInfo $file): string
	{
		$data = 0;

		if (!file_exists($this->dir))
			mkdir($this->dir, 0777, true);

		$filename = $file->getPathname()[0] != '.' ? $file->getPathname() : substr($file->getPathname(), 1);
		$filename = $this->dir . str_replace(DIRECTORY_SEPARATOR, '', $filename);

		if (!file_exists($filename)) {
			$fileHandle = fopen($filename, 'w');
			if (!$fileHandle)
				return 0;

			fwrite($fileHandle, '0');
			fclose($fileHandle);
		}

		$fileHandle = fopen($filename, 'r');
		if (!$fileHandle)
			return 0;

		$data = fread($fileHandle, (int) filesize($filename));
		fclose($fileHandle);
		return strval($data);
	}
}
```
